internal/proxyservice/sshagent/proxy_service.go

```diff
@@ -0,0 +1,165 @@
+package sshagent
+
+import (
+	"crypto/x509"
+	"encoding/pem"
+	"fmt"
+	"io"
+	"net"
+	"strconv"
+
+	validation "github.com/go-ozzo/ozzo-validation"
+	"golang.org/x/crypto/ssh/agent"
+
+	"github.com/cyberark/secretless-broker/internal"
+	"github.com/cyberark/secretless-broker/pkg/secretless/log"
+)
+
+type proxyService struct {
+	done                bool
+	listener            net.Listener
+	logger              log.Logger
+	keyring				agent.Agent
+	retrieveCredentials internal.CredentialsRetriever
+}
+
+// NewProxyService constructs a new instance of a SSH Agent ProxyService. The
+// constructor takes a Listener, Logger and CredentialResolver.
+// An SSH Agent ProxyService is able to server SSH Agent requests with Credentials
+func NewProxyService(
+	listener net.Listener,
+	logger log.Logger,
+	retrieveCredentials internal.CredentialsRetriever,
+) (internal.Service, error) {
+	errors := validation.Errors{}
+
+	if retrieveCredentials == nil {
+		errors["retrieveCredentials"] = fmt.Errorf("retrieveCredentials cannot be nil")
+	}
+	if listener == nil {
+		errors["logger"] = fmt.Errorf("logger cannot be nil")
+	}
+	if listener == nil {
+		errors["listener"] = fmt.Errorf("listener cannot be nil")
+	}
+
+	if err := errors.Filter(); err != nil {
+		return nil, err
+	}
+
+	return &proxyService{
+		retrieveCredentials: retrieveCredentials,
+		listener:            listener,
+		logger:              logger,
+		done:                false,
+		keyring: 			  agent.NewKeyring(),
+	}, nil
+}
+
+func parseKey(pemStr []byte) (rawkey interface{}, err error) {
```

Comment:
nit: This func might look better with unnamed params since then you can return x509.Parse... methods as-is